### PR TITLE
chore(core): reduce default number of WAL apply threads

### PR DESCRIPTION
**Pull request by puzpuzpuz**
The default number of WAL threads seems too generous. Having 2 threads for the smallest boxes is too much and, in small transaction cases, having fewer WAL apply threads leads to better throughput.

### CODE REVIEW

**Comment by puzpuzpuz** [comment]
@ideoma could you take a look at the changes and let me know what you think?

**Comment by ideoma** [comment]
#### [PR Coverage check]

:disappointed: **fail** : 2 / 5 (40.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|2|5|40.00%|

File: core/src/main/java/io/questdb/PropServerConfiguration.java [hunk 1]
```diff
@@ -660,17 +660,17 @@ public PropServerConfiguration(
...
+            cpuWalApplyWorkers = 2;
         } else if (cpuAvailable > 16) {
-            cpuWalApplyWorkers = 4;
+            cpuWalApplyWorkers = 2;
             cpuSpare = 1;
             // tested on 4/32/48 core servers
             cpuIoWorkers = cpuAvailable / 2;
         } else if (cpuAvailable > 32) {
-            cpuWalApplyWorkers = 4;
+            cpuWalApplyWorkers = 3;
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/PropServerConfiguration.java:673]: It's also 3 on 48 core machine. We can add an extra breaking point maybe for the 4 value, e.g. 64+ cores?

[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/PropServerConfiguration.java:673]: how about 25%?

[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/PropServerConfiguration.java:673]: 25% of 32 is 8 threads - seems too many for the WAL apply, no? Maybe we should simply keep 4 threads on this line?

[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/PropServerConfiguration.java]: I reverted to 4 threads as the default for this branch in e073e6b978db881832a2e5baa4d4d90cd783091f
PTAL

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 4 / 6 (66.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|4|6|66.67%|

**Comment by puzpuzpuz** [comment]
@ideoma I've pushed a test enhancement discussed [here](https://github.com/questdb/questdb/pull/4876#discussion_r1722703161) to this PR. See d823fea39ebc85d8f99ec359327c3e6b618d9c02

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 8 / 10 (80.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|4|4|100.00%|

File: core/src/main/java/io/questdb/PropServerConfiguration.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/PropServerConfiguration.java]: >  in small transaction cases, having fewer WAL apply threads leads to better throughput
What about medium to large transactions? 
I agree, maybe on the smallest boxes this PR adds some value but overall, on 8+ CPU cores why writing to 2 tables simultaneously is on average better than writing into 3 tables?

[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/PropServerConfiguration.java]: > on 8+ CPU cores why writing to 2 tables simultaneously is on average better than writing into 3 tables? What's the intuition?

The intuition is that with more tables than threads more transactions are ready to be applied when one of the threads starts processing WAL segments for a table. When doing the recent 3.5M row/s experiment, I had to decrease WAL apply pool to make it less than the default.

**Review by ideoma — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 8 / 10 (80.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|4|4|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 8 / 10 (80.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|4|6|66.67%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|4|4|100.00%|

**Pull request merged**